spatialdata/_core/aggregate.py

```diff
@@ -0,0 +1,164 @@
+from __future__ import annotations
+
+import anndata as ad
+import dask.dataframe as ddf
+import geopandas as gpd
+import numpy as np
+import pandas as pd
+from scipy import sparse
+
+import spatialdata as sd
+from spatialdata._core.models import PointsModel, ShapesModel, get_schema
+
+
+def aggregate(
+    values: ddf.DataFrame | gpd.GeoDataFrame,
+    by: gpd.GeoDataFrame,
+    id_key: str,
+    *,
+    value_key: str | None = None,
+    agg_func: str = "mean",
+) -> ad.AnnData:
+    """
+    Aggregate values by given shapes
+
+    Parameters
+    ----------
+    values
+        Values to aggregate. Currently, only supports Points or Shapes.
+    by
+        Regions to aggregate by. Currently, only supports Shapes.
+    id_key
+        Key to group observations in `values` by. E.g. this could be transcript id for points.
+    value_key
+        Key to aggregate values by. This is the key in the values object.
+        If nothing is passed here, assumed to be a column of ones.
+
+        For points, this could be probe intensity.
+    agg_func
+        Aggregation function to apply over point values, e.g. "mean", "sum", "count".
+        Passed to pandas.DataFrame.groupby.agg.
+
+    Returns
+    -------
+    AnnData of shape (by.shape[0], values[id_key].nunique())])
+    """
+    # TODO: Check that values are in the same space
```

Comment:
Each element could be associated to multiple spaces. I would add `target_coordinate_system` parameter (like done for the query APIs), and then use the spatial alignment defined in that space. See for instance here: https://github.com/scverse/spatialdata/blob/eb99aebbdb4f7b7ee69d9663c5c996f5fa8f0440/spatialdata/_core/_spatial_query.py#L119

Reply:
I think we may need `target_coordinate_system` due to the current implementation of coordinate systems, but don't think this argument is actually needed, or at least shouldn't be required.

* I think the aggregation should occur in `value`'s intrinsic coordinate space. This is because this function broadly assumes there are more `value`s than `by`'s to group by.
* But right now we can't actually uniquely identify the intrinsic coordinate system for coordinate based values (shapes, points) https://github.com/scverse/spatialdata/issues/169

Reply:
In the [Xenium + Visium example](https://drive.google.com/file/d/11sZz2zOoK4bqR8Dy32oAzFSh0gg40tws/view?usp=share_link), having the aggregation happening in the "intrinsic coordinate space" would be incorrect.

You can see this in the GIF after the text "The two images are now aligned in the aligned coordinate system."

Reply:
Both Visium and Xenium data (which live in different spatialdata objects), map to the "global" coordinate system. But the data is not aligned in this coordinate system. The data is only aligned in the "aligned" coordinate system.

Reply:
Yes, it would be incorrect if the intent was to have two different coordinate spaces, but they were both called `"global"`. But, also, that seems like an extremely error prone thing to do even if we're requiring explicit target spaces.

Reply:
I see, ok this is what happens: the `visium` and `cosmx` functions from the `spatialdata-io` library returns the data (that "lives in separate spaces") and gives the default name "global".

Reply:
The `library_id` value could be reasonable defaults values for 10x coordinate space names, since we broadly assume these are unique anyways.

Reply:
@giovp wdyt? Do you think it's possible to find a library id for each tech (not just visium and xenium)?

Reply:
yeah it would be possible but only for xenium and visium afaik. I think it's more sensible to have a `default` cs that is called something like `"{element_name}_default"`.

@ivirshup regarding this
> But right now we can't actually uniquely identify the intrinsic coordinate system for coordinate based values (shapes, points)

also related to the github issue. I think this is not true, as the intrinsic is the one where we have `Identity`. I don't think there will ever be a case where there are 2 cs with both identity, but we could also just validate that if `Identity` is there, then there can be only one.